商品出品機能

kzen23 opened the pull request:
# What
商品出品機能(ユーザーが出品した商品情報をデータベースに保存)

# Why
商品出品機能を実装するため

・ログイン時、商品出品ページへ遷移できる。
https://gyazo.com/e65e2f56abe74c7daaff150289eb7f1d

・価格を入力すると自動的に販売手数料・販売利益も表示される。
https://gyazo.com/f02d98de804691fde742b07fefa49c9c

・必要な情報を適切に入力して、出品するボタンを押すと商品情報がデータベースに保存される。
その１(出品まで)
https://gyazo.com/11dcf9cf53f98080a249f882b982966b
その２(出品後のデータベースの状態)
https://gyazo.com/d7e50167aae7eb02bff36a3247a6cf9d

・入力に問題がある状態で出品するボタンを押すと情報は保存されず、出品ページへ戻り、エラーメッセージが表示される。
その１(入力に問題がある状態で出品するをクリック)
https://gyazo.com/f298b72b120c3b785b766f108378a672
その２(その後のデータベース)
https://gyazo.com/1f1b9abd6e04610e75433e7e30be6d42

・ログアウト状態で商品出品ページへ遷移しようとするとログインページへ遷移させられる
https://gyazo.com/606e12101fa7660adc2e9d8601954d53

・モデル単体テストコード結果
https://gyazo.com/ec2e0d655b6f10f2315952350979433b

kzen23 committed:
itemモデルとitemsテーブル作成

kzen23 committed:
アソシエーション

userモデルとitemモデル

kzen23 committed:
ActiveStorage導入

has_one_attached :image記述済み

kzen23 committed:
Active Hash導入・作成・記述

kzen23 committed:
バリデーションを設定

kzen23 committed:
商品出品ページを表示できるようにする

kzen23 committed:
出品した商品情報をテーブルに保存できるようにする

kzen23 committed:
itemモデルの単体テストコード

kzen23 committed:
必要なJSファイルを作成とテスト

kzen23 committed:
必要な要素の取得・テスト

priceの入力フォームのid情報を取得し、ちゃんと取得できているかを確かめた。

kzen23 committed:
入力するたびにイベント発火するように

inputイベントを仕込み、動作確認、成功。

kzen23 committed:
必要なvalue属性を取得できるように

kzen23 committed:
JavaScriptで販売手数料・利益計算結果の表示

ユーザーの出品価格入力に応じて販売手数料・利益計算結果を表示できるようにした。

kzen23 committed:
rubocopによる自動修正

shodai-ito commented on app/controllers/application_controller.rb:
```diff
@@ -1,17 +1,18 @@
 class ApplicationController < ActionController::Base
   before_action :basic_auth
-  # before_action :authenticate_user!
+  before_action :authenticate_user!, except: [:index]
```
`before_action :authenticate_user!, except: [:index]`
上記記述はItemsControllerに記述しましょう。
ApplicationContorollerは全てのコントローラーの元であるため、こちらに記述してしまうと他のすべてのコントローラーで挙動してしまいます。

作成するコントローラーが一つだけのシンプルなアプリなら問題ありませんが、今回実装いただくフリマアプリは最終的に複数のコントローラーが必要になるため、適用させたいコントローラーごとに分けてメソッドを適用させる必要があります。

shodai-ito commented on app/javascript/item_price.js:
```diff
@@ -0,0 +1,11 @@
+    const profitDom = document.getElementById("profit");
+    const addTax = inputValue * 0.1
+    addTaxDom.innerHTML = Math.floor(addTax)
+    profitDom.innerHTML = Math.floor(inputValue - addTax)
```
現在の記述ですと、333円のような金額になった際、利益と手数料の合計が合わないかと存じます。
正しい金額が表示されるよう、修正しましょう！

shodai-ito commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,107 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Prefecture can't be blank")
+      end
+      it 'priceの値は半角数字でないと保存できない' do
+        @item.price = '１０００'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is invalid. Input half width characters')
+      end
+      it 'priceの値が299以下だと保存できない' do
+        @item.price = 299
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is out of setting range')
+      end
+      it 'priceの値が1,000,0000以上だと保存できない' do
+        @item.price = 10_000_000
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is out of setting range')
+      end
```
金額については下限値を下回った場合もテストしましょう！

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

その後、再度レビュー依頼フォームのご提出をよろしくお願いいたします。

kzen23 committed:
修正１

ApplicationController内の記述をItemsControllerに移した。

kzen23 committed:
修正2

addTaxをaddTaxValueに変更。

前回の記述では、addTaxは小数点込みの値となってしまうため、Math.floorで計算した後の値をaddTaxValueに代入した。

innerHtmlで挿入する値について前回、記述より簡略化した。

kzen23 committed:
修正3

priceの値に関して、バリデーションで設定した下限値、上限値を利用した形に修正。

kzen23 commented:
修正完了致しました。修正３に関しては、私がメンターの意図したところを見えていない可能性がありますが、よろしくお願い致します。

・販売手数料と販売利益の動画
https://gyazo.com/96caf1dccccdf636ff43573b1b75c6f0

・テスト成功の動画を改めて
https://gyazo.com/7c5caa680e4a7eeafa8832343fa45ba4

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、
ローカルのMasterにPullした後に次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
